---
title: "TOG RNA-seq Workshop 2021: Part 2" 
author: Nikita Telkar 
date: July 2021
output: 
  html_document: 
    keep_md: yes 
    toc: true  
    toc_depth: 4
    toc_float: 
      collapsed: false 
      smooth_scroll: true
    theme: flatly  
    highlight: pygments 
--- 

### 0.0 Introduction

So, in Part 1 of the workshop, we worked through our raw data to make it fit for downstream analysis. *This type of data is called as processed/clean data - and is the most common data type made available through GEO.*  

**However, when submitting/releasing your data publicly, it is *always* good practice to make sure that you release both your raw and processed data, as researchers might want to apply their own analysis methods right from the start to the raw dataset. When only processed data is available, it is difficult to use it for any comparisons (e.g., with your own data or as part of a meta-analysis) as some of the upstream data cleanup steps that have already been applied might differ.**    

***  

```{r, echo = F}
library(formatR)
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80))
```  

Loading packages:

```{r libraries, warning = FALSE, error = FALSE, message = FALSE}

library(tidyverse)
library(here) 
library(rmarkdown)
library(knitr)

#formatting packages
library(kableExtra)
library(janitor)
library(scales)
library(ggpubr)

``` 

Loading our processed data from Part 1:  

```{r loading eNorm}

eNorm <- read.delim(here::here("data", "eNorm.txt"), sep = "\t")
eNorm <- eNorm %>% 
  column_to_rownames(var = "gene")

pDat <- read.delim(here::here("data", "GSE157103_formatted_pDat.txt"), sep = "\t")
pDat <- pDat %>% 
  column_to_rownames(var = "ID")

```


### 6.0 PCA  

We'll start with principle component analysis (PCA) - a dimensionality reduction method that accounts for sample variation while maximizing variance.  

```{r pca}

#transforming eNorm values to log2(x)+1
e_log2 <- log2(eNorm + 1)

#transposing our log(x+1) transposed data frame, so that the columns become the rows, and the rows become columns. As we want to check the variance driven by the genes, and not the samples, we transpose the dataframe to have the columns as the samples, and the rows as the genes, as the PCA function performs column-wise applications, not row-wise.
t_log2 <- as.data.frame(t(e_log2))
# DON'T OPEN this df, as it might crash R
dim(t_log2)

#As our data has already been normalized, we don't want to scale it further. We do however, want to centre it - meaning standardizing the upper and lower limits of the distribution of our values
pca <- prcomp(t_log2, scale = FALSE, center = TRUE)

summary(pca)
screeplot(pca)

# -- COPY - PASTE --
#dataframe with all PCs, their variance, and cumulative variance of all PCs
summary <- data.frame(PC = 1:126, var_explained = (pca$sdev)^2 / sum((pca$sdev)^2), 
                      cumulative = cumsum(pca$sdev^2 / sum(pca$sdev^2))
                      )
summary <- summary %>% 
  mutate(cumulative_perc = cumulative*100)

# -- --

#usually we only consider the first 30 PCs
summary <- summary[1:30,]

#different ways to represent the same data
summary %>%
  ggplot(aes(x = sort(as.factor(PC)), y = var_explained)) +
  geom_bar(stat = "identity", fill = "forest green") +
  # geom_text(aes(label = round(var_explained, digits = 2), vjust = -0.8), size = 2) +
  theme_minimal() +
  labs(title = "Variance Explained by each PC") 

summary %>%
  ggplot(aes(x = sort(as.factor(PC)), y = var_explained))+
  geom_point(colour = "forest green") +
  geom_line(group = "PC", colour = "forest green") +
  theme_minimal() +
  labs(title = "Variance Explained by each PC") 

summary %>%
  ggplot(aes(x = sort(as.factor(PC)), y = cumulative_perc))+
  geom_point(colour = "forest green") +
  geom_line(group = "PC", colour = "forest green") +
  theme_minimal() +
  labs(title = "Cumulative Proportion of Variation") 

#separating the PCA values into its won separate df
scores <- as.data.frame(pca$x)

scores <- scores[c(1:30)]
head(scores)

#making a metadata df containing all sample information data
mDat <- cbind(pDat, scores)

```  

Now that we have our PC scores, we'll estimate which of our variables are the ones driving that variation in our data 

```{r PCA-correaltion}

#remotes::install_github("wvictor14/plomics")
library(plomics)

#here, we'll select the Pvalue metric
variable_variance <- lmmatrix(dep = scores, ind = pDat[c(2:15)], metric = "Pvalue")

head(variable_variance)

vv_plot <- variable_variance %>% 
  as.data.frame() 
vv_plot <- as.data.frame(t(vv_plot))
vv_plot <- vv_plot %>% 
  mutate(Principle_Component = 1:30) %>% 
  dplyr::select(Principle_Component, everything())
head(vv_plot)
vv_plot <- vv_plot %>% 
  pivot_longer(cols = -c(Principle_Component), names_to = "variables", values_to = "pval") 
vv_plot <- vv_plot %>% 
  mutate(pval_cat = case_when(
    pval > 0.05  ~ "> 0.05",
    pval < 0.05 & pval > 0.01 ~ "< 0.05",
    pval < 0.01 & pval > 0.001 ~ "< 0.01",
    pval < 0.001 ~ "< 0.001"
  ))

vv_plot %>% 
  ggplot(aes(x = Principle_Component, y = variables, fill = pval_cat)) +
  geom_tile() + 
  theme_bw() +
  labs(x = "PC", y = "Variables" , fill = "P value")

vv_colpal <- c("< 0.001" = "#ef6a4c", "< 0.01" = "#f59e72", "< 0.05" = "#fde0c5", "> 0.05" = "white")
vv_plot$Principle_Component <- as.factor(vv_plot$Principle_Component)

g3 <- vv_plot %>% 
  ggplot(aes(x = Principle_Component, y = variables, fill = pval_cat)) +
  geom_tile(col = "lightgrey") + 
  theme_bw() +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_manual(values = vv_colpal) +
  coord_fixed() + 
  # theme(legend.position = "bottom") +
  labs(x = "PC", y = "Variables" , fill = "P value")

# ASSIGNMENT 1: Repeat the lmmatrix function, but this time with using the Rsqaured value. Store it as a separate variable, and plot both the pval and rsquared plots one below each other. 
# Instead of using `scale_fill_manual` for the plot, use `scale_fill_gradient(low = "", high = "#")` 

rsq <- lmmatrix(dep = scores, ind = pDat[c(2:15)], metric = "Rsquared")

rsq <- rsq %>% 
  as.data.frame() 

rsq <- as.data.frame(t(rsq))

rsq <- rsq %>% 
  mutate(Principle_Component = 1:30) %>% 
  dplyr::select(Principle_Component, everything())

head(rsq)

rsq <- rsq %>% 
  pivot_longer(cols = -c(Principle_Component), names_to = "variables", values_to = "rsq") 

rsq$Principle_Component <- as.factor(rsq$Principle_Component)

str(rsq)

g4 <- rsq %>% 
  ggplot(aes(x = Principle_Component, y = variables, fill = rsq)) +
  geom_tile(col = "lightgrey") + 
  theme_bw() +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_gradient(low = "white", high = "#ef6a4c") +
  coord_fixed() + 
  # theme(legend.position = "bottom") +
  labs(x = "PC", y = "Variables" , fill = "R-square")

ggarrange(g3, g4, ncol = 1, nrow = 2)

```

We'll now plot the first 2 PCs with the variables that seem to be contributing to the most variance in the data.

```{r pca plots}

mDat %>% 
  ggplot(aes(x = PC1, y = PC2, colour = COVID)) +
  geom_point(size = 3) +
  # coord_cartesian(ylim = c(-130, 130), xlim = c(-130, 130)) +
  labs( x = "Principle Component 1", y = "Principle Component 2", title = "COVID: PC1 vs PC2") +
  scale_colour_manual(values = c("orange", "gray")) +
  theme_minimal() 

mDat %>% 
  ggplot(aes(x = PC1, y = PC2, colour = ICU)) +
  geom_point(size = 3) +
  coord_cartesian(ylim = c(-130, 130), xlim = c(-130, 130)) +
  labs( x = "Principle Component 1", y = "Principle Component 2", title = "ICU: PC1 vs PC2") +
  scale_colour_manual(values = c("grey", "blue")) +
  theme_minimal() 

mDat %>% 
  ggplot(aes(x = PC1, y = PC2, colour = Mechanical_Ventilation)) +
  geom_point(size = 3) +
  coord_cartesian(ylim = c(-130, 130), xlim = c(-130, 130)) +
  labs( x = "Principle Component 1", y = "Principle Component 2", title = "Mechanical Ventilation: PC1 vs PC2") +
  scale_colour_manual(values = c("grey", "purple")) +
  theme_minimal()

mDat %>% 
  # mutate(AP_score = case_when(
  #   APACHEII_Score <= 10 ~ "less_than_10",
  #   between(APACHEII_Score, 11, 20) ~ "eleven_to_20",
  #   between(APACHEII_Score, 21, 30) ~ "twentyone_to_30",
  #   between(APACHEII_Score, 31, 40) ~ "thirtyone_to_40",
  #   APACHEII_Score > 40 ~ "more_than_40")) %>% 
  ggplot(aes(x = PC1, y = PC2, colour = APACHEII_Score)) +
  geom_point(size = 3) +
  coord_cartesian(ylim = c(-130, 130), xlim = c(-130, 130)) +
  labs( x = "Principle Component 1", y = "Principle Component 2", title = "APACHEII_Score", subtitle = "Score of disease-severity measured upon admittance to ICU") +
  theme_minimal() 

#EXTRA - COMPARE PC2 and PC3
#PC2-3

mDat %>% 
  ggplot(aes(x = PC2, y = PC3, colour = Ventilator_free_days)) +
  geom_point(size = 3) +
  coord_cartesian(ylim = c(-100, 100), xlim = c(-100, 100)) +
  labs( x = "Principle Component 1", y = "Principle Component 2", title = "Ventilator Free Days : PC1 vs PC2") +
  #scale_colour_manual(values = c("grey", "purple")) +
  theme_minimal()

mDat %>% 
  ggplot(aes(x = PC2, y = PC3, colour = ICU)) +
  geom_point(size = 3) +
  coord_cartesian(ylim = c(-100, 100), xlim = c(-100, 100)) +
  labs( x = "Principle Component 1", y = "Principle Component 2", title = "ICU: PC1 vs PC2") +
  scale_colour_manual(values = c("grey", "blue")) +
  theme_minimal() 

```  

### 7.0 Differential Expression Analysis  

Okay, now moving into DE analysis: we're going to use the limma package, rather than the more popular DESeq2 or edgeR packages. There's broadly 3 steps to pulling out DE genes:  

1. Specifying your variables of interest to generate a model in the form of a matrix    
2. Fitting our data to that model
3. Applying Bayesian statistics to the results of our model

```{r DE covid}

library(limma)

mm_covid <- model.matrix(~COVID, pDat) 
#always better to use an intercept, as the starting value is not forced to zero

head(mm_covid)

all(rownames(pDat) == colnames(eNorm))

efit_COVID <- lmFit(eNorm, mm_covid)

efit_COVID <- efit_COVID %>% 
  eBayes()

#"BH", "BY" and "holm"
topTable(efit_COVID, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05)
topTable(efit_COVID, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05)
topTable(efit_COVID, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, sort.by = "p")
topTable(efit_COVID, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, sort.by = "logFC")

#google S100A9, GBGT1, and COVID - what do we find?

```

We know from our PCA that age doesn't seem to contribute to the variation observed. Let's check whether controlling for age in our model changes the results we obtained  

```{r DE covid age}

mm_age <- model.matrix(~COVID + Age, pDat)

#We'll first, however, get some statistics on the quality of our model with including age
mm_age <- as.data.frame(mm_age)

#logistic requires categorical to be either yes or no
model1 <- glm(COVIDyes ~ Age, data = mm_age, family = binomial)
summary(model1)

#Here the summary shows that age does not seem to strongly correlate with COVID status, and so hence we would not expect a major change in our results on including it in our model (and so hence, we should not). However, just to test that, let's add it to out model and check the resutls.

mm_age <- model.matrix(~COVID + Age, pDat) 

efit_age <- lmFit(eNorm, mm_age) %>% 
  eBayes()

topTable(efit_age, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, sort.by = "logFC")
topTable(efit_age, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, sort.by = "p")

#We see that when arranged by logFC and by adjusted pvalue our model with and without age shows the same ordering of the genes.
#
```  

We saw that lactate concentration was contributing to PC2. Let's check if we should be adjusting for this variable. 

```{r DE covid age lactate}

mm_lactate <- model.matrix(~COVID + Age + Lactate_mmol.l , pDat) 

mm_lactate_df <- as.data.frame(mm_lactate) 

lactate_logres <- glm(COVIDyes ~ Lactate_mmol.l, data = mm_lactate_df, family = binomial)
summary(lactate_logres)

#The summary shows that lactate conc indeed does seem to be significantly associated with COVID status. Let's visualise that
mm_lactate_df %>%
  ggplot(aes(x = Lactate_mmol.l, y = COVIDyes)) +
  geom_point(alpha = 0.2, colour = "orange") +
  geom_smooth(method = "glm", method.args = list(family = "binomial"), colour = "orange") +
  theme_minimal() +
  labs(title = "Does lactate concentration inform of COVID status?", x = "Lactate (mmol/l)", y = "Probability of COVID-positive status")

#so now we know that there is a significant association with lactate levels and the probability of having COVID. Let's add lactate to our linear model
efit_lactate <- lmFit(eNorm, mm_lactate) %>% 
  eBayes()

topTable(efit_lactate, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, sort.by = "logFC")
topTable(efit_lactate, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, sort.by = "p")


# ASSIGNMENT 2: calculate if any 2 variables of your choice might have a correlation with COVID status by using the geom_smooth argument in ggplot

pDat %>% 
  mutate(covid = case_when(
    COVID == "yes" ~ 1,
    COVID == "no" ~ 0)) %>% 
  ggplot(aes(x = CRP_mg.l, y = covid)) +
  geom_point(alpha = 0.2, colour = "orange") +
  geom_smooth(method = "glm", method.args = list(family = "binomial"), colour = "orange") +
  theme_minimal() +
  labs(title = "Does CRP concentration inform of COVID status?", x = "Lactate (mmol/l)", y = "Probability of COVID-positive status")

``` 

Given that we saw that the S100A9 gene showed the highest negative logFC, let's compare the expression of S100A9 between COVID positive nad negative patients.  

```{r S100A9}

S100A9 <- eNorm %>% 
  rownames_to_column(var = "gene") %>% 
  filter(gene == "S100A9") %>% 
  column_to_rownames(var = "gene")
  
S100A9 <- as.data.frame(t(S100A9))

S100A9 <- S100A9 %>% 
  rownames_to_column(var = "sample")

pDat <- pDat %>% 
  rownames_to_column(var = "sample")

covid <- pDat %>% 
  dplyr::select(sample, COVID)
  
S100A9 <- S100A9 %>% 
  left_join(covid, by = "sample")

S100A9 %>% 
  ggplot(aes(x = COVID, y = log2(S100A9), fill = COVID)) +
  geom_boxplot() +
  scale_fill_manual(values = c("gray", "orange")) +
  theme_minimal() + 
  theme(legend.position = "bottom") +
  labs(x = "COVID Status", y = "log2 (S100A9 RPM)", title = "S100A9: Gene with highest negative logFC change")

#The difference doesn't seem much because S100A9 was pulled out from a very simple model where we only included COVID status as a variable.

```

```{r GBGT1 and HBA2}

# GBGT1 <- eNorm %>% 
#   rownames_to_column(var = "gene") %>% 
#   filter(gene == "GBGT1") %>% 
#   column_to_rownames(var = "gene")
#   
# GBGT1 <- as.data.frame(t(GBGT1))
# 
# GBGT1 <- GBGT1 %>% 
#   rownames_to_column(var = "sample")
#   
# GBGT1 <- GBGT1 %>% 
#   left_join(covid, by = "sample")
# 
# GBGT1 %>% 
#   ggplot(aes(x = COVID, y = log2(GBGT1), fill = COVID)) +
#   geom_violin() +
#   scale_fill_manual(values = c("gray", "orange")) +
#   theme_minimal() + 
#   theme(legend.position = "bottom") +
#   labs(x = "COVID Status", y = "log2 (GBGT1 RPM)", title = "GBGT1: Gene with lowest adjusted p-value with or without accounting for lactate")
# 


#ASSIGNMENT 3: do the same plots for HBA2 but use geom_violin to visualize expression and facet_wrap to separate by COVID status

HBA2 <- eNorm %>% 
  rownames_to_column(var = "gene") %>% 
  filter(gene == "HBA2") %>% 
  column_to_rownames(var = "gene")
  
HBA2 <- as.data.frame(t(HBA2))

HBA2 <- HBA2 %>% 
  rownames_to_column(var = "sample")
  
HBA2 <- HBA2 %>% 
  left_join(covid, by = "sample")

HBA2 %>% 
  ggplot(aes(x = COVID, y = log2(HBA2), fill = COVID)) +
  geom_violin() +
  geom_dotplot(binaxis = "y", stackdir = "center", dotsize = 0.7, fill = "black") +
  scale_fill_manual(values = c("gray", "orange")) +
  theme_minimal() + 
  theme(legend.position = "bottom") +
  labs(x = "COVID Status", y = "log2 (HBA2 RPM)", title = "HBA2: Gene with highest negative logFC change on including lactate concentration in the model")

# HBA2 %>% 
#   ggplot(aes(x = sample, y = log2(HBA2), colour = COVID)) +
#   geom_point() +
#   scale_colour_manual(values = c("gray", "orange")) +
#   theme_minimal() + 
#   theme(legend.position = "bottom") +
#   labs(x = "COVID Status", y = "log2 (HBA2 RPM)", title = "HBA2: Gene with highest negative logFC change on including lactate concentration in the model") +
#   facet_grid(~COVID)

```

In case of wanting to measure the differences in the expression between timepoints, we'll use a function called as `makeContrasts` We're make a dummy protein to work through the process:

```{r timepoint makecontrasts}

## -- COPY - PASTE ---

set.seed(500)

#we'll call this dummy protein `unicorn`

unicorn_COVID <- pDat %>% 
  filter(COVID == "yes") %>% 
  dplyr::select(sample, COVID)

#specifying number of obvs, mean, and sd
unicorn_COVID <- unicorn_COVID %>% 
  mutate(unicorn_0days = rnorm(n = 100, mean = 1000, sd = 20), 
         unicorn_7days = rnorm(n = 100, mean = 500, sd = 20), 
         unicorn_14days = rnorm(n = 100, mean = 100, sd = 20))

unicorn_nonCOVID <- pDat %>% 
  filter(COVID == "no") %>% 
  dplyr::select(sample, COVID)

unicorn_nonCOVID <- unicorn_nonCOVID %>% 
  mutate(unicorn_0days = rnorm(n = 26, mean = 100, sd = 10), 
         unicorn_7days = rnorm(n = 26, mean = 100, sd = 10), 
         unicorn_14days = rnorm(n = 26, mean = 100, sd = 10))

pDat_unicorn <- rbind(unicorn_COVID, unicorn_nonCOVID)

pDat_unicorn <- pDat_unicorn %>% 
  dplyr::select(-COVID)

pDat <- pDat %>% 
  right_join(pDat_unicorn, by = "sample")

unicorn <- pDat %>% 
  dplyr::select(sample, COVID, unicorn_0days, unicorn_7days, unicorn_14days)

unicorn <- unicorn %>% 
  pivot_longer(cols = 3:5, names_to = "days", values_to = "measurment")

# unicorn %>% 
#   ggplot(aes(x = days, y = measurment, fill = days)) +
#   geom_boxplot() +
#   scale_fill_manual(values = c("maroon", "hotpink2", "pink")) +
#   theme_minimal() + 
#   theme(legend.position = "bottom") +
#   labs(x = "Day at Measurement", y = "Measurement (in units)", title = "Measurement of Unicorn Protein over Days") +
#   facet_wrap(~COVID)

unicorn$days <- fct_relevel(unicorn$days, c("unicorn_0days", "unicorn_7days", "unicorn_14days"))

unicorn %>% 
  ggplot(aes(x = days, y = measurment, fill = days)) +
  geom_boxplot() +
  scale_fill_manual(values = c("maroon", "hotpink2", "pink")) +
  theme_minimal() + 
  theme(legend.position = "bottom") +
  labs(x = "Day at Measurement", y = "Measurement (in units)", title = "Measurement of Unicorn Protein over Days") +
  facet_wrap(~COVID)

## ---

mm_unicorn <- model.matrix(~COVID + unicorn_0days + unicorn_7days + unicorn_14days, pDat) 

mm_unicorn <- as.data.frame(mm_unicorn)
model2 <- glm(COVIDyes ~ unicorn_0days + unicorn_7days + unicorn_14days, data = mm_unicorn, family = binomial) 
summary(model2)

#we know that there's no actual protein called `unicorn` and hence it isn't actually associated with any changed in gene expression - as we can see from our linear model summary

lmfit_unicorn <- lmFit(eNorm, mm_unicorn)

conmat_unicorns <- makeContrasts(
  day14_day7 = unicorn_14days - unicorn_7days,
  day7_day0 = unicorn_7days - unicorn_0days,
  day14_day0 = unicorn_14days - unicorn_0days,
  levels = mm_unicorn
)

#intercept colname is different
rownames(conmat_unicorns) = colnames(mm_unicorn)

confit_unicorn <- contrasts.fit(lmfit_unicorn, conmat_unicorns) %>% 
  eBayes()

topTable(confit_unicorn)

decideTests(confit_unicorn, adjust.method = "fdr", p.value = 0.05) %>%
  summary()

```  

Let's now select all of the DEG that we got when we included lactate measurement in out model  

### 8.0 Enrichment Analysis

```{r GO enrichment}

#BiocManager::install("biomaRt")

library(biomaRt)

listMarts() #gives us the list of databases available
ensembl=useMart("ensembl")

#which all species are present from the ensembl database?
head(listDatasets(ensembl))

#let's put all of the species into one variable in order to view
biomart_dfs <- as.data.frame(listDatasets(ensembl)$dataset)
head(biomart_dfs)

#we'll make a variable selecting the Homo sapiens dataset
mart = useMart("ensembl", dataset="hsapiens_gene_ensembl")

# Using the DEGs we got from the lactate model
genes <- topTable(efit_lactate, coef = "COVIDyes", adjust.method = "fdr", p.value = 0.05, n = Inf, sort.by = "logFC")
genes <- rownames(genes)

head(genes)

filters <- listFilters(mart) #filters are the parameters you search by
attr <- listAttributes(mart) #attributes are the matching parameters you're wanting to search for

#we require the Entrz IDs for all functions after this step - so converting HGNC Symbols to Entrez IDs
hgnc_to_entrez <- getBM(attributes = c("hgnc_symbol", "entrezgene_id"), filters = "hgnc_symbol", values = genes, mart = mart)

head(hgnc_to_entrez)

#selecting attributes as the GO id, the GO term, the GO term definition, and the cell comparatment that GO term belongs to, searching by the filter/parameter HGNC symbol
go_terms <- getBM(attributes = c("hgnc_symbol", "go_id", "name_1006", "definition_1006", "namespace_1003"), filters = "hgnc_symbol", values = genes, mart = mart)

head(go_terms)

#deleting all empty rows
go_terms <- go_terms %>% 
  mutate_all(na_if,"")
go_terms <- na.omit(go_terms)

#counting the frequency of each GO term
go_plot <- go_terms %>% 
  dplyr::count(name_1006) %>% 
  dplyr::arrange(desc(n))

#we know that the total DEGs were 5134, so let's get the percentage of how many of the genes were associated with a particular  GO Term
head(go_plot)
go_plot$total <- 5134
go_plot <- go_plot[-1,]
go_plot <- go_plot %>% 
  mutate(perc = (n/total)*100) %>% 
  dplyr::arrange()

head(go_plot)

#for the first 20 GO Terms
go_plot[1:20,] %>% 
  ggplot(aes(x = name_1006, y = perc)) +
  geom_bar(stat = "identity", width = 0.6)

go_plot[1:20,] %>% 
  ggplot(aes(x = reorder(name_1006, -perc), y = perc)) +
  geom_bar(stat = "identity", width = 0.6) +
  coord_cartesian(y = c(0,100)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Top 20 GO Terms", subtitle = "For DEGs at adjpval <= 0.05", x = "GO Term", y = "Percentage of DEGs assoc. with GO Term")

#let's all add the cellular compartment to our plot
component <- go_terms %>% 
  dplyr::select(name_1006, namespace_1003) %>% 
  distinct(name_1006, .keep_all = TRUE)

head(component)

go_plot <- go_plot %>% 
  right_join(component, by = "name_1006")

head(go_plot)

```

```{r GOplot, fig.height=7}

go_plot[1:20,] %>% 
  ggplot(aes(x = reorder(name_1006, -perc), y = perc, fill = namespace_1003)) +
  geom_bar(stat = "identity", width = 0.6) +
  scale_fill_manual(values = c("maroon", "navy", "forest green")) +
  coord_cartesian(y = c(0,100)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "bottom") +
  scale_y_continuous(breaks = seq(0, 100, 10)) +
  labs(title = "Top 20 GO Terms", subtitle = "For DEGs at adjpval <= 0.05", x = "GO Term", y = "Percentage of DEGs assoc. with GO Term")

```  

Let's look at the pathways associated with our DEGs   

```{r KEGG}

library(clusterProfiler)

k <- enrichKEGG(gene = genes, organism = "hsa")
#aah, so when we input our genes as HGNC IDs, the function doesn't work. We'll use our Entrez IDs that we have

head(hgnc_to_entrez)

k <- enrichKEGG(gene = hgnc_to_entrez$entrezgene_id, organism = "hsa")
head(k)

```

